speech_to_text/speech_to_text.livemd

```diff
@@ -0,0 +1,217 @@
+    {[stream_format: {:output, %Membrane.RemoteStream{}}], %{state | chunk_size: chunk_size}}
+  end
+
+  @impl true
+  def handle_buffer(:input, buffer, _ctx, state) do
+    input = state.queue <> buffer.payload
+
+    if byte_size(input) > @vad_chunk_size do
+      process_data(input, %{state | queue: <<>>})
+    else
+      {[], %{state | queue: input}}
+    end
+  end
+
+  defp process_data(data, state) do
+    # Here we filter out the silence at the beginning of each chunk.
+    # This way we can fit as much speech in a single chunk as possible
+    # and potentially remove whole silent chunks, which cause
+    # model hallucinations. If after removing the silence the chunk
+    # is not empty but too small to process, we store it in the state
+    # and prepend it to the subsequent chunk.
+    speech =
+      if state.speech == <<>> do
+        filter_silence(data, state)
+      else
+        state.speech <> data
+      end
+
+    if byte_size(speech) < state.chunk_size do
+      {[], %{state | speech: speech}}
+    else
+      model_input = Nx.from_binary(speech, :f32)
+      result = Nx.Serving.run(state.serving, model_input)
+      transcription = Enum.map_join(result.chunks, & &1.text)
+      buffer = %Membrane.Buffer{payload: transcription}
+      {[buffer: {:output, buffer}], %{state | speech: <<>>}}
+    end
+  end
+
+  defp filter_silence(samples, state) do
+    samples
+    |> generate_chunks()
+    |> Enum.drop_while(&(calc_volume(&1) < state.vad_threshold))
+    |> Enum.join()
+  end
+
+  defp generate_chunks(samples) when byte_size(samples) >= 2 * @vad_chunk_size do
+    <<chunk::binary-size(@vad_chunk_size), rest::binary>> = samples
+    [chunk | generate_chunks(rest)]
+  end
+
+  defp generate_chunks(samples) do
+    [samples]
+  end
+
+  defp calc_volume(chunk) do
+    samples = for <<sample::float-32-little <- chunk>>, do: sample
+    samples_cnt = Enum.count(samples)
+    samples_avg = Enum.sum(samples) / samples_cnt
+    sum_mean_square = samples |> Enum.map(&((&1 - samples_avg) ** 2)) |> Enum.sum()
+    :math.sqrt(sum_mean_square / samples_cnt)
+  end
+end
+```
+
+## Pipeline structure definition
+
+The pipeline consists of the following elements:
+
+* The `Membrane.PortAudio.Source` - responsible for fetching the audio input from your microphone
+* The `SpeechToText` sink we have previously created
+* The `Membrane.Debug.Sink` that will print the transcriptions to the standard output
+
+```elixir
+import Membrane.ChildrenSpec
+alias Membrane.RCPipeline
+
+spec =
+  child(%Membrane.PortAudio.Source{channels: 1, sample_format: :f32le, sample_rate: 16_000})
+  |> child(SpeechToText)
+  |> child(%Membrane.Debug.Sink{handle_buffer: &IO.puts(&1.payload)})
+
+pipeline = RCPipeline.start_link!()
+RCPipeline.exec_actions(pipeline, spec: spec)
+```
+
+The prints with the transcription should start appearing below the last cell you have evaluated.
```

Comment:
#How about warning the user that he might be asked for permissions to use microphone?